File: challenge-2/Perfect.java
/*
* Java Program to check whether input is a perfect
* number or not
* Input Constrained between 3 and 4999 (inclusive)
*/

import java.util.Scanner;

public class Perfect {

	private int number_to_check;

	// Returns true if number_to_check is perfect
	private boolean isPerfect() {

		int sum_of_divisors = 0;

		// Iterate from 1 to exactly half of the number
		for (int i = 1; i <= (int) (number_to_check / 2); i++) {
			if (number_to_check % i == 0) {
				sum_of_divisors += i;
			}
		}

		if (number_to_check == sum_of_divisors) {
			return true;
		}
		else
		{
			return false;
		}
	}

	// Driver code
	public static void main(String[] args) {
		// Define Scanner object for taking input
		Scanner scanner = new Scanner(System.in);

		// Class object, because static methods can't access
		// non-static method directly.
		Perfect perfect = new Perfect();

		// Iterate and prompt until the constraint
		// is satisfied.
		do {
			System.out.print("Input: ");
			perfect.number_to_check = scanner.nextInt();
		} while (perfect.number_to_check <= 2 || perfect.number_to_check >= 5000);

		if (perfect.isPerfect()) {
			System.out.println(perfect.number_to_check + " is a perfect number");
		} else {
			System.out.println(perfect.number_to_check + " is not a perfect number");
		}
	}
}

File: challenge-2/perfect.c
/*
* C Program to find out whether a given number
* is perfect or not. A perfect number is one in which
* sum of the proper divisors of the number results
* in the number itself. 
* Input Constraint: Greater than 2, less than 5000
*/

#include <stdio.h>

// Returns true if a number is perfect
int is_perfect(int number);

// Driver code
int main()
{
	int number_to_check;
	// Keep taking input until constraint satisfied
	do
	{
		printf("Input: ");
		scanf("%d", &number_to_check);
	} while (number_to_check <= 2 || number_to_check >= 5000);

	// is_perfect returns 1 if argument is perfect
	if (is_perfect(number_to_check))
	{
		printf("%d is a perfect number\n", number_to_check);
	}
	else
	{
		printf("%d is not a perfect number\n", number_to_check);
	}
}

int is_perfect(int number)
{
	int sum_of_divisors = 0;
	// Iterate till half of the parameter to find
	// all proper factors and add them together
	for (int i = 1; i <= number / 2; i++)
	{
		if (number % i == 0)
		{
			sum_of_divisors += i;
		}
	}

	// No comment needed, self explanatory
	if (sum_of_divisors == number)
	{
		return 1;
	}
	else
	{
		return 0;
	}
}
